src/Data/Text/Internal.hs

```diff
@@ -47,18 +49,20 @@ module Data.Text.Internal
 import Control.Exception (assert)
 import GHC.Stack (HasCallStack)
 #endif
+import Control.Monad.ST (ST)
 import Data.Bits
 import Data.Int (Int32, Int64)
 import Data.Text.Internal.Unsafe.Char (ord)
 import Data.Typeable (Typeable)
 import qualified Data.Text.Array as A
+import qualified Language.Haskell.TH.Syntax as TH
 
 -- | A space efficient, packed, unboxed Unicode text type.
 data Text = Text
```

Comment:
May I suggest changing the data constructor name (e.g. `TextUtf8`)? I think you'd really like to break any code using the constructor.

`hashable`, for instance, happily compiles into a buffer over-read

```
instance Hashable T.Text where
    hashWithSalt salt (T.Text arr off len) =
        hashByteArrayWithSalt (TA.aBA arr) (off `shiftL` 1) (len `shiftL` 1)  -- OOPS!
        salt
    ....
hashByteArrayWithSalt ba !off !len !h =
    fromIntegral $ c_hashByteArray ba (fromIntegral off) (fromIntegral len)
    (fromIntegral h)
```

Reply:
> I think you'd really like to break any code using the constructor.

Not quite, code using `Text` constructor could in general retain its semantics. It's unwrapping of `Data.Text.Array.Array`, which causes issues here. I renamed its constructor to highlight the issue for downstream dependencies, thanks!